tensorly/parafac2_tensor.py

```diff
@@ -0,0 +1,511 @@
+        return message
+
+
+def _validate_parafac2_tensor(parafac2_tensor):
+    """Validates a parafac2_tensor in the form (weights, factors)
+    
+        Returns the rank and shape of the validated tensor
+    
+    Parameters
+    ----------
+    parafac2_tensor : Parafac2Tensor or (weights, factors)
+    
+    Returns
+    -------
+    (shape, rank) : (int tuple, int)
+        size of the full tensor and rank of the Kruskal tensor
+    """
+    if isinstance(parafac2_tensor, Parafac2Tensor):
+        # it's already been validated at creation
+        return parafac2_tensor.shape, parafac2_tensor.rank
+
+    weights, factors, projections = parafac2_tensor
+            
+    if len(factors) != 3:
+        raise ValueError('A PARAFAC2 tensor should be composed of exactly three factors.'
+                         'However, {} factors was given.'.format(len(factors)))
+
+    if len(projections) != factors[0].shape[0]:
+        raise ValueError('A PARAFAC2 tensor should have one projection matrix for each horisontal'
+                         ' slice. However, {} projection matrices was given and the first mode has'
+                         'length {}'.format(len(projections), factors[0].shape[0]))
+
+    rank = int(T.shape(factors[0])[1])
+
+    shape = []
+    for i, projection in enumerate(projections):
+        current_mode_size, current_rank = T.shape(projection)
+        if current_rank != rank:
+            raise ValueError(
+                'All the projection matrices of a PARAFAC2 tensor should have the same number of '
+                'columns as the rank. However, rank={} but projections[{}].shape[1]={}'.format(
+                    rank, i, T.shape(projection)[1]
+                )
+            )
+
+        inner_product = T.dot(T.transpose(projection), projection)
+        if T.max(T.abs(inner_product - T.eye(rank))) > 1e-5:
+            raise ValueError(
+                'All the projection matrices must be orthonormal, that is, P.T@P = I. '
+                'However, T.norm(projection[{}].T@projection[{}] - T.eye(rank)) = {}'.format(
+                    i, i, T.norm(inner_product - T.eye(rank)) 
+                )
+            )
+        
+        shape.append((current_mode_size, *[f.shape[0] for f in factors[2:]]))  # Tuple unpacking to possibly support higher order PARAFAC2 tensors in the future
+
+    
+    # Skip first factor matrix since the rank is extracted from it.
+    for i, factor in enumerate(factors[1:]):
+        current_mode_size, current_rank = T.shape(factor)
+        if current_rank != rank:
+            raise ValueError('All the factors of a PARAFAC2 tensor should have the same number of columns.'
+                             'However, factors[0].shape[1]={} but factors[{}].shape[1]={}.'.format(
+                                 rank, i, T.shape(factor)[1]))
+
+    if weights is not None and len(weights) != rank:
```

Comment:
No idea why tensorflow doesn't support this. I guess we can do.

```suggestion
    if weights is not None and tl.shape(weights)[0]  != rank:
```